### PR TITLE
Update ws-directive build toolchain to webpack, npm.

### PR DESCRIPTION
**Pull request by SethDavenport**
_Sorry for the enormous PR. Most of the changes are due to the addition of `package-lock.json` and the removal of `bower_components` from github._

# Toolchain Upgrade

This PR updates the ws-directives build and dependency management tools, specifically:

* gulp has been replaced with webpack
* dependencies specified in bower have been moved to npm as peerDependencies

As part of this, the way we specify JS modules and the way CSS and templates get included in the output has changed (more below).

# Rationale

This is part of an ongoing effort to address some weaknesses in the gulp/bower toolchain across our web stack.

## Issues with bower

* bower is deprecated. People aren't publishing things to it as much any more. A recent example was the onboarding document signing component - we want to leverage a 3rd party dependency that's only on npm. Now we can.
* the mobile team uses npm. If we want to be able to share code with them, we need to at least be on the same dependency registry.
* bower doesn't give us much control over dependency collision resolution or shrinkwrapping. This poses a risk of breaking deployments.

## Issues with Gulp

* Gulp is from the days before JS had a real module syntax. Code is full of IIFEs for isolation, and files have to be concatenated together in the right order or they break. It's also doesn't have a good model for avoiding unused libraries or processing non-JS files (HTML, SCSS, etc.).

## Issues with modularity

* angular.module was a good idea before JS had modules. It was the only way to resolve dependencies between chunks of application code. However it means that business logic code is all the more tightly coupled to AngularJS, again limiting our ability to share code with the mobile team.

# How does this affect consumers of ws-directives?

TL;DR: It doesn't. We still publish ws-directives to bower as before. We still publish ws-directives.js and ws-directives.css the same way as before, and they contain the same stuff.

We do publish source maps now too, which is an added bonus.

# How this this affect people developing ws-directives?

There are a few changes to the way modularization is handled.

## Defining an angular module has changed

Instead of scattering `angular.module('foo')` calls in each of your files, we now create an `index.js` file for the angular module you want to define. For example, `ws.common.filters`:

```js
import angular from 'angular';

// Our app code can be implemented as regular functions and imported here.
import ChunkFilterFactory from './chunk.filter';
import RemoveSalutationFilterFactory from './remove-salutation.filter';
import TitleCaseFilterFactory from './title-case.filter';

// We export the name of our module so that the parent module can consume it.
const injectorName = 'ws.common.filters';
export default injectorName;

// We wire up all the parts of the module in one spot.
angular.module(injectorName, [])
  .filter('chunk', ChunkFilterFactory)
  .filter('removeSalutation', RemoveSalutationFilterFactory)
  .filter('titleCase', TitleCaseFilterFactory);
```

The parent module can just import and use it like this:

```js
import angular from 'angular';

import filterModule from './filters';
// ...

const injectorName = 'ws.common';
export default injectorName;

angular.module(injectorName, [
  filterModule,
  // ...
])
// ...
```

In effect, we're separating the calls to `angular.module.*` into a "wirings file" whose job is to tell Angular how to assemble our application at runtime.

The rest of the code is just plain-old-JS.

## Creating an angular component should be simpler

This means that a component is now a lot simpler:

```js
// Just import templates directly - no more $templateCache hackery. Webpack transforms it
// for you with (html-loader).
import template from './expander.html';

// Does this component use some CSS? you could import it here:
import './expander.scss';

const ExpanderComponent = {
  template,
  controller: 'wsExpanderController',
  bindings: {
    showText: '@?',
    hideText: '@?',
    expanded: '<?',
    onExpand: '&?',
    onCollapse: '&?',
  },
  transclude: {
    header: '?wsExpanderHeader',
    body: 'wsExpanderBody',
  },
};

// Just exporting a regular object.
export default ExpanderComponent;
```

Note that we don't need 'use strict' or the IIFE any more, since ES6 modules are strict and isolated by default.

## You can now use classes more effectively if you want

I've enabled a couple of babel options that let us define controllers as classes if we want:

```js
export default class ExpanderController {
  // What an injection annotation looks like in a class:
  static $inject = ['$log'];

  // Stuff gets injected into the constructor.
  constructor($log) {
    $log.info('A message');
    this.showText = this.showText || '+';
    this.hideText = this.hideText || '-';
    this.expanded = this.expanded || false;
  }

  toggle = () => {
    this.expanded = !this.expanded;
    if (this.expanded && this.onExpand) {
      this.onExpand();
    }

    if (!this.expanded && this.onCollapse) {
      this.onCollapse();
    }
  }

  getToggleText = () =>
    this.expanded ?
      this.hideText :
      this.showText;

  isExpanded = () => this.expanded;
}
```

## Unit testing should be simpler

Note that the controller above has nothing angular-specific about it. We're passing dependencies as constructor arguments (_dependency inversion_) and letting the module file tell Angular how to assemble stuff at runtime (_dependency injection_). But our tests no longer need all that `inject(` stuff because `ExpanderController` is just a class with some constructor args:

```js
import ExpanderController from './expander.controller';

describe('wsExpanderController', () => {
  let controller;
  let mock$log;

  beforeEach(() => {
    mock$log = { info: () => null; };
    spyOn(mock$log, 'info');

    controller = new ExpanderController(mock$log);
  });

  it('logs a message when created', () => {
    expect(mock$log.info).toHaveBeenCalled();
  });

  it('changes the button text when toggled', () => {
    controller.showText = 'ohai';
    controller.hideText = 'kthxbai';

    expect(controller.getToggleText()).toBe('ohai');

    controller.toggle();
    expect(controller.getToggleText()).toBe('kthxbai');

    controller.toggle();
    expect(controller.getToggleText()).toBe('ohai');
  });
});
```

Note that if you are writing tests that involve compiling templates, you'll still need the angular mocks module stuff, but the syntax is slightly different: `angular.mock.module(...)` and `angular.mock.inject(...)`:

```js
import commonModule from '../..';

describe('wsFooter directive', () => {
  let outerScope;
  let isolatedScope;
  let element;

  beforeEach(angular.mock.module(commonModule));

  function injectElement(elementHtml) {
    beforeEach(angular.mock.inject(($rootScope, $compile) => {
      outerScope = $rootScope.$new();
      element = angular.element(elementHtml);
      $compile(element)(outerScope);
      outerScope.$digest();
      isolatedScope = element.isolateScope();
    }));
  }

  // ...
}
```

# Future improvements

This lib no longer uses bower for its dependencies, but we still publish it to bower. We now also publish it to npm using the name `@wealthsimple/angular-directives`.

Webpack's powerful loader model opens the door to CSS improvements such as autoprefixer, scoping. Or other potentially interesting things like HAML templates (!).

### CODE REVIEW

**Comment by SethDavenport** [comment]
Pro tip: you can cut down on whitespace change noise by adding `?w=1` to the end of the PR URL, e.g.: https://github.com/wealthsimple/ws-directives/pull/72/files?w=1

**Comment by SethDavenport** [comment]
Related to the following:

* https://trello.com/c/BDKUu6fM/13-ws-directives-try-out-the-share-owner-admin-approach-to-angular
* https://trello.com/c/z2ty3xFU/4-ws-directives-publish-to-npm
* https://trello.com/c/d0wIIWWX/12-ws-directives-linter-doesnt-check-test-files

**Comment by SethDavenport** [comment]
@DanGodfrey @evancloutier as discussed on Friday. Should now be straighforward to include the signning module. Come bug me for help once it's merged.

**Comment by SethDavenport** [comment]
I'll need someone with admin repo access to adjust the GH pages config once this is merged. I can show you what needs doing.

**Comment by SethDavenport** [comment]
yeah that's a good idea - I'll see if I can put one together.
